bin/utils.js

```diff
@@ -227,11 +228,18 @@ const pingTimeout = 30000
  * @param {any} req
  * @param {any} opts
  */
-exports.setupWSConnection = (conn, req, { docName = req.url.slice(1).split('?')[0], gc = true } = {}) => {
+exports.setupWSConnection = async (conn, req, { docName = req.url.slice(1).split('?')[0], gc = true } = {}) => {
   conn.binaryType = 'arraybuffer'
   // get doc, initialize if it does not exist yet
   const doc = getYDoc(docName, gc)
   doc.conns.set(conn, new Set())
+
+  // await the doc state being updated from persistence, if available, otherwise
+  // we may send sync step 2 too early
+  if (doc.whenSynced) {
+    await doc.whenSynced
+  }
```

Comment:
Note: If I move this await below the message listeners the original problem persists